### PR TITLE
allow installing of localgov_workflows from config import

### PR DESCRIPTION
**Pull request by finnlewis**
Recreate pull request from https://github.com/Polynya https://github.com/localgovdrupal/localgov_workflows/pull/64 to allow installing of localgov_workflows from config import, where assumed dependencies might not be available at install time.

## What does this change?

Changes the way the localgov_workflows atemtps to add the localgov_editorial workflow to all new content  types.

## How to test

Good question! 

Here's one way to generate an error in a standard install, without this branch.

1. Install a new localgov_drupal site
2. Export database backup
3. Enable localgov_workflows and localgov_guides
4. Export config
5. Import db 
6. Import config

Or more specifically: 

1. install localgov profile

```
ddev drush sql-dump > db.sql
ddev drush en localgov_workflows localgov_guides
ddev drush cex -y
ddev drush sql-drop -y
ddev drush sqlc < db.sql -y
ddev drush cim -y
```

See this error: 

```
 [notice] Synchronized extensions: install workflows.
 [notice] Synchronized extensions: install content_moderation.
 [notice] Synchronized extensions: install diff.
 [notice] Synchronized extensions: install localgov_guides.
 [notice] Synchronized extensions: install responsive_preview.
 [notice] Synchronized extensions: install scheduled_transitions.

In EntityType.php line 916:
                                                                                     
  Missing bundle entity, entity type node_type, entity id localgov_guides_overview.  
                                                                                     
```

## Have we considered potential risks?

After this change, some content types might not get the localog_editorial workflow added automagically.

### CODE REVIEW

**Comment by ekes** [comment]
My comments from the previous MR stand. It's moving hook_install into an update hook. These are two completely different things. One runs when you install it, and it won't run any existing update hooks. The other runs the first time it sees the hook after it was installed.

The hook_install should use the $is_syncing and bail out like we do elsewhere (if it doesn't already.

The entity insert hook. That's more interesting. It certainly should check if the workflow exists, but with this module installed it should exist. So it again would be nice to do is_syncing (because in that case we shouldn't tinker with the workflows, the imported ones should be used). But that's not something on offer there, is it? Or is there a clue on the entity itself?

**Comment by finnlewis** [comment]
Following @ekes's line of thinking, if we are synching the config and installing the module at that point, I think we don't want to do anything in hook_install.

I could be mistaken, I often am!

If we've installed the module locally, then exported the config, why would we need to do anything on hook_install again?

So I've gone with simply: 

```php
  if ($is_syncing) {
    // If syncing config, don't run the install process.
    return;
  }
```
Thoughts @stephen-cox @ekes ?

File: localgov_workflows.install [hunk 1]
```diff
@@ -18,10 +18,16 @@ use Drupal\workflows\Entity\Workflow;
...
+    // If syncing config, don't run the install process.
+    return;
+  }
+
   // Configure workflow for localgov_ node bundles with no other workflow.
   $editorial = Workflow::load('localgov_editorial');
   if (!$editorial) {
     $config_key = 'workflows.workflow.localgov_editorial';
+    $editorial_config = FALSE;
     if ($is_syncing) {
```
Comment (ekes) [inline localgov_workflows.install]: This will want removing now as it won't ever get here if `$is_syncing`

**Review by ekes — COMMENTED** [review]
To me it makes sense that if it's syncing it shouldn't need to configure the workflow or scheduled transitions or roles at all. That should have been done on the site that the user is syncing from where the module will have been enabled. So the sync should have all these already configured.

If the end user has changed something - like removed it from a content type maybe - we have to assume they intended to do this.

**Comment by finnlewis** [comment]
@stephen-cox I'd love to get your input on this. 
@Polynya I'd also love to know if this solves your issue too from pull request  https://github.com/localgovdrupal/localgov_workflows/pull/64 and https://github.com/localgovdrupal/localgov_workflows/issues/63

File: localgov_workflows.install [hunk 2]
```diff
@@ -43,8 +39,9 @@ function localgov_workflows_install($is_syncing) {
   $node_types = \Drupal::service('entity_type.bundle.info')->getBundleInfo('node');
   $changed = FALSE;
   foreach ($node_types as $type_name => $node_type) {
+    // For content types named localgov_* that don't have a workflow, add the
+    // localgov_editorial workflow.
     if (strpos($type_name, 'localgov_') === 0 && empty($node_type['workflow'])) {
-
       // Add workflow.
```
Comment (stephen-cox) [inline localgov_workflows.install:45]: Nitpick - this comment is superfluous.

**Review by stephen-cox — APPROVED** [review]
This looks good to me :+1: 

The install hook only makes config changes so it shouldn't do anything if syncing.

It also fixes installs from existing config as this now works:

```
drush si localgov -y
drush localgov_workflows localgov_guides -y
drush cex -y
drush si localgov --existing-config -y
```

Although you'll either need to manually delete the `install_localgov` hook from the profile or patch core, see https://www.drupal.org/project/drupal/issues/2982052

**Comment by finnlewis** [comment]
Thanks @stephen-cox !

**Comment by Polynya** [comment]
Thanks @finnlewis . I haven't had chance to test it but looks good to me.

**Pull request merged**